modules/telestion-api/src/main/java/de/wuespace/telestion/api/message/JsonMessage.java

```diff
@@ -1,141 +1,346 @@
+	 * when the conversion completes successfully or fails when a {@link DecodeException} occurs during conversion.
 	 *
-	 * @param json source which must be a String
-	 * @param type class of the message
-	 * @param <T>  type of the {@link Message}
-	 * @return decoded message object
+	 * @return a future that represents the conversion state
 	 */
-	static <T extends JsonMessage> T from(Object json, Class<T> type) {
-		return JSON_CODEC.fromString((String) json, type);
+	static <T extends JsonMessage> Future<T> on(Class<T> type, Message<?> message) {
+		return Future.future(promise -> on(type, message, promise::complete, promise::fail));
 	}
 
+	///
+	/// SYNCHRONOUS DECODING SECTION
+	///
+
 	/**
-	 * Returns the simple class name of the subclass.
+	 * Constructs a {@link JsonMessage} from a buffer which contains an encoded JSON string.
 	 *
-	 * @return simple class name of subclass
+	 * @param json the buffer that contents contain the necessary information to construct
+	 *             the specified {@link JsonMessage}
+	 * @param type the class of the target {@link JsonMessage}
+	 * @param <T>  the type of the target {@link JsonMessage}
+	 * @return the decoded message
+	 * @throws DecodeException if the buffer contents does not contain the necessary information to successfully
+	 *                         construct the specified {@link JsonMessage}
 	 */
-	@JsonProperty(access = JsonProperty.Access.READ_ONLY)
-	default String className() {
-		return getClass().getName();
+	static <T extends JsonMessage> T from(Buffer json, Class<T> type) throws DecodeException {
+		return Json.CODEC.fromBuffer(json, type);
 	}
 
 	/**
-	 * Returns the Json-representation of the message.
+	 * Constructs a {@link JsonMessage} from a JSON {@link String}.
 	 *
-	 * @return json representation of message
+	 * @param json the JSON {@link String} that contains the necessary information to construct
+	 *             the specified {@link JsonMessage}
+	 * @param type the class of the target {@link JsonMessage}
+	 * @param <T>  the type of the target {@link JsonMessage}
+	 * @return the decoded message
+	 * @throws DecodeException if the JSON string does not contain the necessary information to successfully
+	 *                         construct the specified {@link JsonMessage}
 	 */
-	default JsonObject json() {
+	static <T extends JsonMessage> T from(String json, Class<T> type) throws DecodeException {
+		return Json.CODEC.fromString(json, type);
+	}
+
+	/**
+	 * Constructs a {@link JsonMessage} from a plain {@link Object}.
+	 *
+	 * @param json the plain {@link Object} that contains the necessary information to construct
+	 *             the specified {@link JsonMessage}
+	 * @param type the class of the target {@link JsonMessage}
+	 * @param <T>  the type of the target {@link JsonMessage}
+	 * @return the decoded message
+	 * @throws DecodeException if the plain object does not contain the necessary information to successfully
```

Comment:
what about the possible IllegalArgumentException which you experienced?

Reply:
Ok, that's an interesting case:
`Json.CODEC.fromValue(json, type)` call in the `DatabindCodec` implementation the `DatabindCodec.mapper.convertValue` method which in turn **can** throw an `IllegalArgumentException`.
But depending on the installed packages, the `JacksonCodec` could also be used which instantly throws a `DecodeException`.